tests/python_package_test/test_basic.py

```diff
@@ -877,3 +877,57 @@ def test_feature_names_are_set_correctly_when_no_feature_names_passed_into_Datas
         data=np.random.randn(100, 3),
     )
     assert ds.construct().feature_name == ["Column_0", "Column_1", "Column_2"]
+
+
+# NOTE: this intentionally contains values where num_leaves <, ==, and > (max_depth^2)
+@pytest.mark.parametrize("max_depth,num_leaves", [(-1, 3), (-1, 50), (5, 3), (5, 31), (5, 32), (8, 3), (8, 31)])
```

Comment:
[nit] I prefer to differentiate between parameters more clearly :smile:

```suggestion
@pytest.mark.parametrize(("max_depth" ,"num_leaves"), [(-1, 3), (-1, 50), (5, 3), (5, 31), (5, 32), (8, 3), (8, 31)])
```

Reply:
sure, I'm indifferent.